fuzz/fuzzers/unbuffered.rs

```diff
@@ -0,0 +1,95 @@
+#![no_main]
+#[macro_use]
+extern crate libfuzzer_sys;
+extern crate rustls;
+
+use std::sync::Arc;
+
+use rustls::client::UnbufferedClientConnection;
+use rustls::server::{ResolvesServerCert, UnbufferedServerConnection};
+use rustls::unbuffered::{ConnectionState, UnbufferedStatus};
+use rustls::{ClientConfig, RootCertStore, ServerConfig, SideData};
+
+fuzz_target!(|data: &[u8]| {
+    let mut data = data.to_vec();
+    match data.split_first_mut() {
+        Some((0x00, rest)) => client(rest),
+        Some((0x01, rest)) => server(rest),
+        Some((_, _)) | None => {}
+    }
+});
+
+fn client(data: &mut [u8]) {
+    let root_store = RootCertStore::empty();
+    let config = ClientConfig::builder()
+        .with_root_certificates(root_store)
+        .with_no_client_auth();
+    let conn =
+        UnbufferedClientConnection::new(config.into(), "localhost".try_into().unwrap()).unwrap();
+    fuzz_unbuffered(data, ClientServer::Client(conn));
+}
+
+fn server(data: &mut [u8]) {
+    let config = ServerConfig::builder()
+        .with_no_client_auth()
+        .with_cert_resolver(Arc::new(Fail));
+    let conn = UnbufferedServerConnection::new(config.into()).unwrap();
+    fuzz_unbuffered(data, ClientServer::Server(conn));
+}
+
+enum ClientServer {
+    Client(UnbufferedClientConnection),
+    Server(UnbufferedServerConnection),
+}
+
+fn fuzz_unbuffered(mut data: &mut [u8], mut conn: ClientServer) {
+    while !data.is_empty() {
+        let consumed = match &mut conn {
+            ClientServer::Server(server) => {
+                let st = server.process_tls_records(data);
+                process(st)
+            }
+            ClientServer::Client(client) => {
+                let st = client.process_tls_records(data);
+                process(st)
+            }
+        };
+
+        match consumed {
+            Some(consumed) => {
+                data = &mut data[consumed..];
+            }
+            None => break,
+        };
+    }
+}
+
+fn process<S: SideData>(status: UnbufferedStatus<'_, '_, S>) -> Option<usize> {
+    let UnbufferedStatus { discard, state } = status;
+
+    match state {
+        Ok(ConnectionState::EncodeTlsData(mut enc)) => {
+            let mut buffer = [0u8; 16384];
```

Comment:
is this buffer size significant in some way? Maybe the literal could be written `16_384`?

Reply:
fixed; should actually be `16_384 + 5`. Maybe in the future we could export this as a const.